ultradns-rest/src/main/java/denominator/ultradns/model/Region.java

```diff
@@ -0,0 +1,238 @@
+
+  private String name;
+  private String code;
+  private String type;
+  private int id;
+  private String effectiveCode;
+  private String effectiveCodeForGeo;
+
+  private Region parentRegion;
+  private TreeSet<Region> childRegions;
+
+  public Region(String name, String code, String type, int id) {
+    this.name = name;
+    this.code = code;
+    this.type = type;
+    this.id = id;
+    this.setEffectiveCode();
+    this.setEffectiveCodeForGeo();
+  }
+
+  public String getName() {
+    return this.name;
+  }
+
+  public void setName(String name1) {
+    this.name = name1;
+  }
+
+  public String getCode() {
+    return this.code;
+  }
+
+  public void setCode(String code1) {
+    this.code = code1;
+  }
+
+  public String getType() {
+    return this.type;
+  }
+
+  public void setType(String type1) {
+    this.type = type1;
+  }
+
+  public int getId() {
+    return this.id;
+  }
+
+  public void setId(int id1) {
+    this.id = id1;
+  }
+
+  public Region getParentRegion() {
+    return this.parentRegion;
+  }
+
+  public void setParent(Region region) {
+    this.parentRegion = region;
+    this.setEffectiveCode();
+    this.setEffectiveCodeForGeo();
+  }
+
+  public TreeSet<Region> getChildRegions() {
+    return this.childRegions;
+  }
+
+  public void setChildRegions(Collection<Region> regions) {
+    this.childRegions = new TreeSet<Region>(regions);
+    for (Region childRegion : childRegions) {
+      childRegion.setParent(this);
+    }
+  }
+
+  public String getEffectiveCode() {
+    if (this.effectiveCode == null && this.code != null) {
+      this.setEffectiveCode();
+    }
+    return this.effectiveCode;
+  }
+
+  public void setEffectiveCode() {
+    if (this.getParentRegion() != null) {
+      this.effectiveCode = this.getParentRegion().getCode() + "-" + this.getCode();
+    } else {
+      this.effectiveCode = this.getCode();
+    }
+  }
+
+  public String[] getEffectiveCodesOfChildRegions() {
+    String[] childRegionEffectiveCodes = new String[this.getChildRegions().size()];
+    int i = 0;
+    for (Region region : this.getChildRegions()) {
+      childRegionEffectiveCodes[i] = region.getEffectiveCode();
+      i++;
+    }
+    Arrays.sort(childRegionEffectiveCodes);
+    return childRegionEffectiveCodes;
+  }
+
+  public String getEffectiveCodeForGeo() {
+    if (this.effectiveCodeForGeo == null && this.code != null) {
+      this.setEffectiveCodeForGeo();
+    }
+    return this.effectiveCodeForGeo;
+  }
+
+  public void setEffectiveCodeForGeo() {
+    if (this.isCountry() || this.isRegion()) {
+      this.effectiveCodeForGeo = this.getCode();
+    } else {
+      String parentRegionCode = "";
+      if (this.getParentRegion() != null) {
+        parentRegionCode = this.getParentRegion().getCode();
+      }
+      this.effectiveCodeForGeo = parentRegionCode + "-" + this.getCode();
+    }
+  }
+
+  public boolean isRegion() {
+    return this.getType().equals("Region");
```

Comment:
Can there be a region which will have a null value for its "type"? If yes, then instead of creating and raising a new exception like "RegionTypeIsNullException", this NullPointerException will be simpler. Please suggest.